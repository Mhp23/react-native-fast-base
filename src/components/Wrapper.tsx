import React from 'react';

import {StyleProp, View, ViewStyle} from 'react-native';
import {WrapperProps} from '../core/types';

const defaultProps: WrapperProps = {
  mode: 'normal',
  flex: false,
};

const Wrapper: React.FC<WrapperProps> = ({
  flex,
  mode,
  style,
  alignX,
  alignY,
  children,
  spaceX,
  spaceY,
  spaceEnd,
  spaceStart,
  spaceTop,
  spaceBottom,
  ...rest
}) => {
  const wrapperStyles: StyleProp<ViewStyle> = {
    justifyContent: alignY,
    alignItems: alignX,
    ...(spaceY
      ? {paddingVertical: spaceY}
      : {
          paddingTop: spaceTop,
          paddingBottom: spaceBottom,
        }),
    ...(spaceX
      ? {paddingHorizontal: spaceX}
      : {
          paddingEnd: spaceEnd,
          paddingStart: spaceStart,
        }),
    ...(flex && {flex: 1}),
    ...(mode !== 'normal' && {
      flexDirection: mode,
      flexWrap: 'wrap',
    }),
  };

  return (
    <View style={[wrapperStyles, style]} {...rest}>
      {children}
    </View>
  );
};

Wrapper.defaultProps = defaultProps;

export default Wrapper;
