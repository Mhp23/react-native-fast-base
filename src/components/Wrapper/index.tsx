import React from 'react';
import {useStyle} from '../../hooks';
import type {WrapperProps} from '../../core';
import {useRM} from 'react-native-full-responsive';
import {colorSelector, makeStyle} from '../../utils';
import {StyleSheet, View, type ViewStyle} from 'react-native';

const defaultProps: WrapperProps = {
  mode: 'normal',
  flex: false,
  p: 0,
  px: 0,
  py: 0,
  pt: 0,
  pb: 0,
  pe: 0,
  ps: 0,
  m: 0,
  mx: 0,
  my: 0,
  mt: 0,
  mb: 0,
  me: 0,
  ms: 0,
};

const Wrapper = React.forwardRef<View, WrapperProps>(
  (
    {
      p,
      px,
      py,
      pt,
      pb,
      pe,
      ps,
      m,
      mx,
      my,
      mt,
      mb,
      me,
      ms,
      self,
      flex,
      mode,
      style,
      width,
      height,
      alignX,
      alignY,
      children,
      background,
      ...rest
    },
    ref,
  ) => {
    const {rs, rh, rw} = useRM();

    const backgroundStyle = useStyle(() => {
      const bgColor = colorSelector(background);
      return bgColor && {backgroundColor: bgColor};
    }, [background]);

    const wrapperStyles = useStyle<ViewStyle>(() => {
      const wrapperFlex = (typeof flex === 'number' && flex) || (flex && 1);
      return makeStyle<ViewStyle>({
        flex: wrapperFlex,
        width: rw(width),
        height: rh(height),
        alignSelf: self,
        alignItems: alignX,
        justifyContent: alignY,
        ...(mode !== 'normal' && {
          flexDirection: mode,
          flexWrap: 'wrap',
        }),
      });
    }, [rw, rh, self, alignX, alignY, width, height, flex, mode]);

    const spacesStyles = useStyle<ViewStyle>(() => {
      return makeStyle<ViewStyle>({
        padding: rs(p),
        paddingEnd: rs(pe),
        paddingTop: rs(pt),
        paddingStart: rs(ps),
        paddingBottom: rs(pb),
        paddingVertical: rs(py),
        paddingHorizontal: rs(px),
        margin: rs(m),
        marginEnd: rs(me),
        marginTop: rs(mt),
        marginStart: rs(ms),
        marginBottom: rs(mb),
        marginVertical: rs(my),
        marginHorizontal: rs(mx),
      });
    }, [rs, p, pe, pt, ps, pb, py, px, m, me, mt, ms, mb, my, mx]);

    return (
      <View
        ref={ref}
        style={StyleSheet.flatten([
          backgroundStyle,
          wrapperStyles,
          spacesStyles,
          style,
        ])}
        testID="FAST_BASE_WRAPPER"
        {...rest}>
        {children}
      </View>
    );
  },
);

Wrapper.defaultProps = defaultProps;

export default Wrapper;
