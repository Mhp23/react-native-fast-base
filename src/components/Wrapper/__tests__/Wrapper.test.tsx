import React from 'react';
import Wrapper from '..';
import {Text, TouchableOpacity} from 'react-native';
import {render, cleanup} from '@testing-library/react-native';

afterEach(cleanup);

const WrppaerID = 'FAST_BASE_WRAPPER';

describe('Wrapper component', () => {
  it('should match snapshot', () => {
    const rendered = render(<Wrapper children={<Text />} />).toJSON();
    expect(rendered).toBeTruthy();
    expect(rendered).toMatchSnapshot();
  });

  it('should be flex equal to 1 when the flex property is true', () => {
    const rendered = render(<Wrapper flex />);
    const wrppaerComponent = rendered.getByTestId(WrppaerID);
    expect(wrppaerComponent.props.style).toMatchObject({flex: 1});
  });

  it('should be flex equal to the passed number', () => {
    let flex = 3;
    const rendered = render(<Wrapper flex={flex} />);
    const wrppaerComponent = rendered.getByTestId(WrppaerID);
    expect(wrppaerComponent.props.style).toMatchObject({flex});
  });

  it('should render row direction mode correctly', () => {
    const rendered = render(
      <Wrapper mode="row">
        <TouchableOpacity testID="FIRST_TOUCHABLE_BUTTON" onPress={() => {}} />
        <TouchableOpacity testID="SECOND_TOUCHABLE_BUTTON" onPress={() => {}} />
      </Wrapper>,
    );
    const wrppaerComponent = rendered.getByTestId(WrppaerID);
    expect(wrppaerComponent.props.style).toMatchObject({flexDirection: 'row'});
    expect(wrppaerComponent.children).not.toBe(undefined);
  });

  it('should render the spaces properties correctly', () => {
    const rendered = render(
      <Wrapper
        p={1}
        px={1}
        py={1}
        pe={1}
        pt={1}
        ps={1}
        pb={1}
        m={1}
        mx={1}
        my={1}
        me={1}
        mt={1}
        ms={1}
        mb={1}
      />,
    ).toJSON();

    const keys = [
      'padding',
      'paddingHorizontal',
      'paddingVertical',
      'paddingEnd',
      'paddingStart',
      'paddingTop',
      'paddingBottom',
      'margin',
      'marginHorizontal',
      'marginVertical',
      'marginEnd',
      'marginStart',
      'marginTop',
      'marginBottom',
    ];
    expect(rendered).toBeTruthy();
    expect(rendered).toMatchSnapshot();
    keys.forEach(styleKey => {
      expect(rendered.props.style).toHaveProperty(styleKey);
    });
  });
});
