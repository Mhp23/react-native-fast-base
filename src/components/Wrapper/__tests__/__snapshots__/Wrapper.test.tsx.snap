// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wrapper component should match snapshot 1`] = `
<View
  style={Object {}}
  testID="FAST_BASE_WRAPPER"
>
  <Text />
</View>
`;

exports[`Wrapper component should render the spaces properties correctly 1`] = `
<View
  style={
    Object {
      "margin": 2.0833333333333335,
      "marginBottom": 2.0833333333333335,
      "marginEnd": 2.0833333333333335,
      "marginHorizontal": 2.0833333333333335,
      "marginStart": 2.0833333333333335,
      "marginTop": 2.0833333333333335,
      "marginVertical": 2.0833333333333335,
      "padding": 2.0833333333333335,
      "paddingBottom": 2.0833333333333335,
      "paddingEnd": 2.0833333333333335,
      "paddingHorizontal": 2.0833333333333335,
      "paddingStart": 2.0833333333333335,
      "paddingTop": 2.0833333333333335,
      "paddingVertical": 2.0833333333333335,
    }
  }
  testID="FAST_BASE_WRAPPER"
/>
`;
